python/tvm/relax/training/optimizer.py

```diff
@@ -0,0 +1,582 @@
+                R.Tuple(R.Tensor((3, 3), dtype="float32")), R.Tuple(R.Tensor((), dtype="int64"))
+            ):
+                with R.dataflow():
+                    lv3: R.Tensor((3, 3), dtype="float32") = params[0]
+                    lv12: R.Tensor((3, 3), dtype="float32") = gradients[0]
+                    lv21: R.Tensor((), dtype="int64") = optim_states[0]
+                    lv31: R.Tensor((), dtype="int64") = R.add(lv21, R.const(1, "int64"))
+                    lv4: R.Tensor((3, 3), dtype="float32") = R.multiply(
+                        R.const(0.1, "float32"), lv12
+                    )
+                    lv5: R.Tensor((3, 3), dtype="float32") = R.subtract(lv3, lv4)
+                    gv1: R.Tuple(R.Tensor((3, 3), dtype="float32")) = (lv5,)
+                    gv11: R.Tuple(R.Tensor((), dtype="int64")) = (lv31,)
+                    R.output(gv1, gv11)
+                return (gv1, gv11)
+        """
+        raise NotImplementedError()
+
+
+def _get_np_shape(var):
+    return [int(val) for val in var.struct_info.shape]
+
+
+def _get_np_dtype(var):
+    return var.struct_info.dtype
+
+
+# We need to subtract on hyperparameters, but do not want to introduce floating point error.
+# Floating point error would lead to a few problems, such as making assert_structural_equal not
+# passed in unit tests
+def _high_precision_subtract(lhs, rhs):
+    return float(Decimal(str(lhs)) - Decimal(str(rhs)))
+
+
+class SGD(Optimizer):
+    """Implements stochastic gradient descent.
+
+    The returned function is equivalent to the following numpy code:
+
+    .. code-block:: python
+        def SGD(param_tuple, grad_tuple, state_tuple):
+            num_steps = state_tuple[0]
+            state_tuple[0] = num_steps + 1
+            for i in range(len(param_tuple)):
+                param = param_tuple[i]
+                grad = grad_tuple[i]
+                param_tuple[i] = param - lr * (grad + weight_decay * param)
+            return param_tuple, state_tuple
+
+    Parameters
+    ----------
+    lr : float
+        learning rate
+
+    weight_decay : Optional[float]
+        weight decay (L2 penalty) (default: 0)
+    """
+
+    def __init__(self, param_list, lr, weight_decay=0):
+        super().__init__(param_list)
+        self.lr = float(lr)
+        self.weight_decay = float(weight_decay)
+
+    @property
+    def state(self):
+        """The state of SGD is `(num_steps,)`.
+
+        Returns
+        -------
+        res : ADT
+            The state of SGD.
+        """
+        if self._state is None:
+            self._state = tuple_object(
+                (
+                    # num_steps = 0
+                    tvm.nd.array(np.zeros((), "int64")),
+                )
+            )
+        return self._state
+
+    @state.setter
+    def state(self, value):
+        self._state = value
+
+    def get_function(self) -> Function:
+        plist = self._param_list
+        len_param = len(plist)
+        dtype = self._dtype
```

Comment:
Is there any reason of defining `plist` and `dtype` here? I see you both use `plist` and `self._param_list` below, and seems that they always refer to the same list.

Reply:
`plist` and `dtype` are just names convenient to use. I have changed all `self._param_list` into `plist`.